feat: add documentation for go sdk (#501)
matter-labs/zksync-web-era-docs: docs/api/go/getting-started.md

# Getting started

In this guide we will demonstrate how to:

1. Connect to the zkSync network.
2. Deposit assets from Ethereum into zkSync.
3. Check balances.
4. Transfer assets.
5. Withdraw assets from zkSync into Ethereum.
6. Deploy a smart contract.
7. Deploy a smart contract with constructor and interact with the contract.
8. Deploy a smart contract with dependency.

## Prerequisites

- Go: >=1.17 ([installation guide](https://go.dev/dl/))

## Adding dependencies

To install zkSync Era, run the command below in your terminal.

```console
 go get github.com/zksync-sdk/zksync2-go
```

## Connecting to zkSync

Once you have all the necessary dependencies, connect to zkSync using the endpoint of the operator node.

```go

> ZkSyncProvider := "https://zksync2-testnet.zksync.dev"

Reviewer: ```suggestion
ZkSyncProvider := "https://testnet.era.zksync.dev"
```

// Connect to zkSync network
zp, err := zksync2.NewDefaultProvider(ZkSyncProvider)
if err != nil {
	log.Panic(err)
}
defer zp.Close()
```

## Examples

Full code for all examples is available [here](https://github.com/zksync-sdk/zksync2-examples/tree/main/go). Examples are configured to interact with `zkSync` and `Goerli` test networks.  

### Deposit funds

This is an example of how to deposit assets from Ethereum network (L1) to zkSync Era network (L2):

```go
package main

import (
	"context"
	"fmt"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/rpc"
	"github.com/zksync-sdk/zksync2-go"
	"log"
	"math/big"
	"os"
)

func main() {
	var (
		PrivateKey       = os.Getenv("PRIVATE_KEY")

> 		ZkSyncProvider   = "https://zksync2-testnet.zksync.dev"

Reviewer: ```suggestion
		ZkSyncProvider   = "https://testnet.era.zksync.dev"
```

	if err != nil {
		log.Panic(err)
	}
	fmt.Println("L2 transaction", l2Hash)

	// Wait for deposit transaction to be finalized on L2 network (5-7 minutes)
	fmt.Println("Waiting for deposit transaction to be finalized on L2 network (5-7 minutes)")
	_, err = zp.WaitMined(context.Background(), l2Hash)
	if err != nil {
		log.Panic(err)
	}

	balance, err = w.GetBalance()
	if err != nil {
		log.Panic(err)
	}
	fmt.Println("Balance after deposit: ", balance)
}

```

### Check balance

Here is an example of how to check a balance on zkSync Era.
```go
func main() {
	var (
		PrivateKey       = os.Getenv("PRIVATE_KEY")
		ZkSyncProvider   = "https://zksync2-testnet.zksync.dev"
	)

	// Connect to zkSync network
	zp, err := zksync2.NewDefaultProvider(ZkSyncProvider)
	if err != nil {
		log.Panic(err)
	}
	defer zp.Close()

	// Create singer object from private key for appropriate chain
	chainID, err := zp.ChainID(context.Background())
	if err != nil {
		log.Fatal(err)
	}
	es, err := zksync2.NewEthSignerFromRawPrivateKey(common.Hex2Bytes(PrivateKey), chainID.Int64())
	if err != nil {
		log.Fatal(err)
	}

	// Create wallet
	w, err := zksync2.NewWallet(es, zp)
	if err != nil {
		log.Panic(err)
	}

>   // Get balance
> 	balance, err := w.GetBalance()
> 	if err != nil {
> 		log.Panic(err)
> 	}

Reviewer: ```suggestion
  // Get balance
	balance, err := w.GetBalance()
	if err != nil {
		log.Panic(err)
	}
	fmt.Println("Balance is: ", balance)
```

	fmt.Println("Finalize withdraw transaction", fwHash)
}

```

### Smart contract deployment

With zkSync, you can deploy a contract using the create method, by simply building the contract into a binary format and deploying it to the zkSync network.

- [Storage](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/storage/Storage.sol): Contract without constructor.
- [Incrementer](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/incrementer/Incrementer.sol): Contract with constructor.
- [Demo](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/demo/Demo.sol): Contract that has a dependency on 
[Foo](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/demo/Foo.sol) contract.

There is a [user guide](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/README.md) on how to compile Solidity smart contracts using `zksolc` 
compiler. `zksolc` compiler generates a `*.zbin` and a `combined.json` file that contains the bytecode and ABI of a smart contract. 
The `combined.json` file is used by `abigen` tool to generate smart contrat bindings. 
Those files are used in the following examples. 

#### Deploy a contract

> ```

Reviewer: ```suggestion
```go
```

In this guide we will demonstrate how to:

1. Connect to the zkSync network.
2. Get all confirmed tokens on zkSync.
3. Deposit assets from Ethereum into zkSync.
4. Check balances.
5. Transfer assets.
6. Withdraw assets from zkSync into Ethereum.
7. Deploy a smart contract.
8. Deploy a smart contract with constructor and interact with the contract.
9. Deploy a smart contract with dependency.

## Prerequisites

- Go: >=1.17 ([installation guide](https://go.dev/dl/))

## Adding dependencies

To install zkSync Era, run the command below in your terminal.

```console
 go get github.com/zksync-sdk/zksync2-go
```

## Connecting to zkSync

Once you have all the necessary dependencies, connect to zkSync using the endpoint of the operator node.

```go
ZkSyncProvider := "https://testnet.era.zksync.dev"

// Connect to zkSync network
zp, err := zksync2.NewDefaultProvider(ZkSyncProvider)
if err != nil {
	log.Panic(err)
}
defer zp.Close()
```

## Examples

Full code for all examples is available [here](https://github.com/zksync-sdk/zksync2-examples/tree/main/go). Examples are configured to interact with `zkSync` and `Goerli` test networks.  

### Get all confirmed tokens on zkSync

> Following example shows how to get all confirmed ERC20 tokens on zkSync network. 

Reviewer: ```suggestion
Following example shows how to get all confirmed tokens on zkSync Era. 
```
